### PR TITLE
sandbox: When testing the reset service, wait for the ledger config.

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
If we don't wait for the ledger configuration, we could potentially ask for completions before the server is fully ready.

This seems to be the source of the flakiness and issues in the reset service tests. I can no longer reproduce the flakes with these changes.

I don't have conclusive proof that these changes make the reset service tests less flaky, but I'm fairly confident. We can revert commit https://github.com/digital-asset/daml/commit/86cfbca7cc2bce9965b920e4d69c6a2d7123e9ef if they turn out to still need some attention.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox-classic/src/test/resources/logback-test.xml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (rautenrieth-da) [inline ledger/sandbox-classic/src/test/resources/logback-test.xml:1]: Is the addition of the new logback configurations intentional?

File: ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala [hunk 2]
```diff
@@ -76,30 +82,43 @@ abstract class ResetServiceITBase
...
+    for {
+      ledgerId <- eventually { (_, _) =>
+        fetchLedgerId()
+      }
+      // Completions won't work until a ledger configuration is in place, so we wait for one.
+      _ <- new StreamConsumer[GetLedgerConfigurationResponse](responseObserver =>
+        LedgerConfigurationServiceGrpc
+          .stub(channel)
+          .getLedgerConfiguration(GetLedgerConfigurationRequest(ledgerId.unwrap), responseObserver))
+        .firstWithin(Span.convertSpanToDuration(scaled(1.second)))
```
Comment (rautenrieth-da) [inline ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala:105]: If I understand correctly, `firstWithin` will complete after a maximum of 1 second, even if the ledger configuration is not ready. Maybe you should check whether the result of `firstWithin` is not an empty vector, and return a failed promise otherwise?

**Review by rautenrieth-da — APPROVED** [review]

File: ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala [hunk 3]
```diff
@@ -10,19 +10,24 @@ import io.grpc.stub.StreamObserver
...
 
   Delayed.by(duration)(onCompleted())
 
   def result: Future[Vector[T]] = promise.future
 
   override def onNext(value: T): Unit = {
     buffer += value
+    size += 1
+    if (maximumResults.exists(size >= _)) {
+      onCompleted()
```
Comment (stefanobaghino-da) [inline ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala:29]: This could possibly fire concurrently with the timer.

1. Is `onCompleted` reentrant?
2. Would it make sense to gate both `onNext` and `onCompleted` on a shared lock (e.g. `promise.synchronized`) and check that the promise is not completed when `onNext` enters the critical section?

File: ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala [hunk 4]
```diff
@@ -10,19 +10,24 @@ import io.grpc.stub.StreamObserver
...
   private val buffer = Vector.newBuilder[T]
+  private var size = 0
 
   Delayed.by(duration)(onCompleted())
 
   def result: Future[Vector[T]] = promise.future
 
   override def onNext(value: T): Unit = {
     buffer += value
+    size += 1
```
Comment (stefanobaghino-da) [inline ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala:27]: `buffer` is a vector builder, it should be fine to just use `buffer.size` instead of keep track of it. It doesn't look like this can cause any problem, but I can see how keeping track of the thing separately can cause weird concurrency bugs in future revisions.

File: ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala [hunk 5]
```diff
@@ -10,19 +10,24 @@ import io.grpc.stub.StreamObserver
 import scala.concurrent.duration.FiniteDuration
 import scala.concurrent.{ExecutionContext, Future, Promise}
 
-final class TimeBoundObserver[T](duration: FiniteDuration)(
+final class TimeBoundObserver[T](duration: FiniteDuration, maximumResults: Option[Int] = None)(
```
Comment (stefanobaghino-da) [inline ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala:13]: Not necessary, but I would recommend adding a secondary constructor that takes an `Int` and wraps it into an option (pure UX).

File: ledger/sandbox/src/test/resources/logback-test.xml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (stefanobaghino-da) [inline ledger/sandbox/src/test/resources/logback-test.xml:1]: I think there should be a shared `logback.xml` that can be used.

File: ledger/sandbox/BUILD.bazel [hunk 7]
```diff
@@ -135,10 +135,7 @@ da_scala_library(
 
 da_scala_test_suite(
     name = "sandbox-tests",
-    srcs = glob(
-        ["src/test/suite/**/*.scala"],
-        exclude = ["src/test/suite/**/ResetService*IT.scala"],
-    ),
+    srcs = glob(["src/test/suite/**/*.scala"]),
```
Comment (stefanobaghino-da) [inline ledger/sandbox/BUILD.bazel:138]: 🙂

File: ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala [hunk 8]
```diff
@@ -184,7 +208,7 @@ abstract class ResetServiceITBase
           .map(ids => ids.distinct should have size 20L)
       }
 
-      // 5 attempts with 5 transactions each seem to strike the right balance to complete before the
+      // 4 attempts with 5 transactions each seem to strike the right balance to complete before the
```
Comment (stefanobaghino-da) [inline ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala:211]: Why this change?

File: ledger/sandbox-classic/src/test/resources/logback-test.xml (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox-classic/src/test/resources/logback-test.xml:1]: Yes. Without them, the logs for the reset service tests (and other tests) are useless. There's way too much debug logging coming from Hikari, Flyway, etc.

File: ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala:105]: Good call. I'll fix that.

File: ledger/sandbox-classic/BUILD.bazel [hunk 9]
```diff
@@ -221,10 +221,7 @@ test_deps = [
 
 da_scala_test_suite(
     name = "sandbox-classic-tests",
-    srcs = glob(
-        ["src/test/suite/**/*.scala"],
-        exclude = ["src/test/suite/**/ResetService*IT.scala"],
-    ),
+    srcs = glob(["src/test/suite/**/*.scala"]),
```
Comment (stefanobaghino-da) [inline ledger/sandbox-classic/BUILD.bazel:224]: 🙂

File: ledger/sandbox-classic/src/test/resources/logback-test.xml (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/sandbox-classic/src/test/resources/logback-test.xml:1]: I'm fairly sure that this applies to all tests that use the shared `logback.xml`, probably that ought to be fixed instead.

File: ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala (hunk 4 above)
Comment (SamirTalwar-DA) [reply ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala:27]: Unfortunately, `buffer.size` is not a thing. The same logic is in `SizeBoundObserver`.

File: ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala (hunk 5 above)
Comment (SamirTalwar-DA) [reply ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala:13]: If we use this more than once, I'd be happy to add it.

File: ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala (hunk 3 above)
Comment (SamirTalwar-DA) [reply ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala:29]: Shouldn't be a problem; `onCompleted` will work idempotently. This logic was copied from `SizeBoundObserver`.

I want to refactor these observers so they're decorators which can be layered on top of each other, but it seems like a big change, so I'll do it separately after this is in.

File: ledger/sandbox/src/test/resources/logback-test.xml (hunk 6 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox/src/test/resources/logback-test.xml:1]: Good call.

Comment (SamirTalwar-DA) [reply ledger/sandbox/src/test/resources/logback-test.xml:1]: Done.

File: ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala:105]: Done.

File: ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala (hunk 8 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox-common/src/test/lib/scala/platform/sandbox/services/reset/ResetServiceITBase.scala:214]: Sorry, forgot to copy this from the commit body.

Now we're also checking for ledger configuration, the time for a reset to be "accepted" is longer, so 5 tries in 30 seconds is a bit flaky. Reducing to 4 seems reasonable.

File: ledger/sandbox-classic/src/test/resources/logback-test.xml (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox-classic/src/test/resources/logback-test.xml:1]: Aha, I see. Somehow, Sadbonx doesn't have a _logback.xml_ file. I didn't realize it was only a problem for those and not Sandbox Classic tests.

I suggest we merge this as-is and revisit Sandbox logging in general soon.

Comment (SamirTalwar-DA) [reply ledger/sandbox-classic/src/test/resources/logback-test.xml:1]: Took another look and I would definitely prefer to keep this separate, because the production Logback configuration for Sandbox doesn't print timestamps. This is very annoying when debugging, because Bazel corrupts the CI timestamps by holding back logs and then dumping them all at once on failure.

It also logs to a file, which is unnecessary for testing.

File: ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala (hunk 3 above)
Comment (SamirTalwar-DA) [reply ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala:29]: Addressed in https://github.com/digital-asset/daml/pull/7984.
